transform/transformers/survey.py

```diff
@@ -0,0 +1,147 @@
+        that survey, and loads it as a Python object.
+
+        :param ids: Survey response ids.
+        :type ids: :py:class:`sdx.common.survey.Survey.Identifiers`
+        :param str package: The name of the Python package where the survey is to be found,
+                            eg: `"sdx.common"`. Within standalone apps which do not declare
+                            themselves as Python packages, pass in the variable `__name__`.
+        :param str pattern: A query for the survey definition. This will be
+                            a file path relative to the package location which uniquely
+                            identifies the survey definition file. It accepts keyword
+                            formatting arguments for any of the attributes of
+                            :py:class:`sdx.common.survey.Survey.Identifiers`.
+
+                            For example: `"surveys/{survey_id}.{inst_id}.json"`.
+        :rtype: dict
+
+        """
+        try:
+            content = pkg_resources.resource_string(
+                package, pattern.format(**ids._asdict())
+            )
+        except FileNotFoundError:
+            return None
+        else:
+            return json.loads(content.decode("utf-8"))
+
+    @staticmethod
+    def bind_logger(log, ids):
+        """Bind a structured logger with survey response metadata.
+
+        :param log: The logger object to be bound.
+        :param ids: The survey response ids to bind to the logger.
+        :type ids: :py:class:`sdx.common.survey.Survey.Identifiers`
+
+        """
+        return log.bind(
+            ru_ref=ids.ru_ref,
+            tx_id=ids.tx_id,
+            user_id=ids.user_id,
+        )
+
+    @staticmethod
+    def parse_timestamp(text):
+        """Parse a text field for a date or timestamp.
+
+        Date and time formats vary across surveys.
+        This method reads those formats.
+
+        :param str text: The date or timestamp value.
+        :rtype: Python date or datetime.
+
+        """
+
+        cls = datetime.datetime
+
+        if text.endswith("Z"):
+            return cls.strptime(text, "%Y-%m-%dT%H:%M:%SZ").replace(
+                tzinfo=datetime.timezone.utc
+            )
+
+        try:
+            return cls.strptime(text, "%Y-%m-%dT%H:%M:%S.%f%z")
+        except ValueError:
+            pass
+
+        try:
+            return cls.strptime(text.partition(".")[0], "%Y-%m-%dT%H:%M:%S")
+        except ValueError:
+            pass
+
+        try:
+            return cls.strptime(text, "%Y-%m-%d").date()
+        except ValueError:
+            pass
+
+        try:
+            return cls.strptime(text, "%d/%m/%Y").date()
+        except ValueError:
+            pass
+
+        if len(text) != 6:
+            return None
+
+        try:
+            return cls.strptime(text + "01", "%Y%m%d").date()
+        except ValueError:
+            return None
```

Comment:
Just a thought while looking at this, is it ever correct for nothing to be passed back by this last try statement?  Seems odd there isn't a single log line in this function that says 'Didn't get enough characters' and that you can pass this function any old rubbish and it'll spit 'something' back (even if that something is nothing) without complaining.

Reply:
I know this is outside the scope of this btw, this can definitely be merged with this in it ^^ If you agree then it might be worth raising a card or a github issue to look at it in future.